### DISCUSSION TITLE
Watch does not work anymore

### DISCUSSION THREAD

**Discussion by wdanilo** [enso-org/enso#6168] [Work In Progress - GUI] [answered]
Hi @PabloBuchu @somebody1234 @indiv0, after I merged develop to my branch (https://github.com/enso-org/enso/commit/0a33cea27f2291c22be6583876a29f5f456f55e9) watch stopped working. After I change rust sources now, the website does not reload automatically anymore as it used to. I assume this is caused by your changes to build scripts. This needs to be fixed asap, as it makes the development much slower. For now, I'm reverting the merge and switching to another branch. The branch from the link above is left for you to see the repro of the issue.

**Comment by somebody1234** [comment]
i probably did break it, not sure how it used to work though, so not sure whether i'll be able to figure out what went wrong
some notes re: esbuild changes:
- i removed the node-based server module completely in favor of esbuild's built-in server
  - however, this should have no effect on `client/`'s `watch` as it uses its own server (`content/` and `dashboard/` use the esbuild one)
- i did also do changes to the esbuild-config scripts, but i don't think i removed anything, only added dashboard builds
- i haven't been able to get `watch` to automatically reload even on old commits
- [this commit](https://github.com/enso-org/enso/commit/ee981d2052e9752b535a26177271be7a1f95ccc7) may be the cause (), however it was merged on march 3, whereas this feature branch was first committed to on march 29, so maybe it's not the cause...
  - it says: "Changes can be pulled by reloading within the electron" - however i can't see anywhere it's automatically reloading so i'm assuming that means manually reloading

**Reply by wdanilo** [reply]
The watch was always (at least on my end) reloading the website automatically after changes. Right now it does not. I do not know if it answers your question of "how it should work", but we need to fix it ASAP (preferably today). If debugging of this issue would take more time, let's revert the commit which caused it. Otherwise, we can get back to the node-based server, or fix it in other way, but we need to make watch working on develop, as it breaks the development speed for many people now.

Regarding the [mentioned commit](https://github.com/enso-org/enso/commit/ee981d2052e9752b535a26177271be7a1f95ccc7), it's definitely not the cause. It was working fine on Mar 27 (based on my git log history). 

Btw, in which PR did you remove the node-based server? Was this PR Q/A tested?

**Comment by PabloBuchu** [comment]
@somebody1234 Did you have chance to relook on this issue after Wojtek comment? Please let us know whats the plan here if you need additional help

**Reply by somebody1234** [reply]
rebuilding an old commit right now, plan is currently to try and figure out how the auto-reload used to work (and why it's broken now)

**Comment by mwu-tow** [comment]
@somebody1234 
> it says: "Changes can be pulled by reloading within the electron" - however i can't see anywhere it's automatically reloading so i'm assuming that means manually reloading

In case of the Electron (`ide watch`), the page is not automatically reloaded; rather the whole Electron application is restarted after the original process exits.
It should however automatically reload when using `gui watch` mode and a web browser. I suspect that https://github.com/enso-org/enso/pull/6034 might have broken this, as it replaced live server with plain esbuild watch.

cc @PabloBuchu

**Reply by somebody1234** [reply]
ah, that makes sense. i believe #6034 does serve the content, however it doesnt add clientside handling to reload the page (an oversight on my part, sorry about that) since i added the live reload to the dashboard entrypoint. so it should be an easy fix, i should be able to implement + test + PR once i get back to my pc

**Comment by PabloBuchu** [comment]
@somebody1234 best way to reproduce this is to open 2 shells:
1. to run `./run gui watch --skip-version-check --skip-wasm-opt`
2. run project manager process `./dist/backend/bin/project-manager`
Then open your browser and navigate to `http://localhost:8080/`

To test reload I am opening `app/gui/view/welcome-screen/src/side_me.rs:32` and change `Your project` to something else.. After watch rebuild the browser page should also automatically reload

**Comment by MichaelMauderer** [comment]
For me, as of today, after merging develop into my working branch, running `./run ide watch --backend-source release --backend-release latest --skip-wasm-opt` no longer seems to start a development server at all or open the electron application. Navigating to `http://localhost:8080/` results in `ERR_CONNECTION_REFUSED`. It works fine for `./run gui watch --skip-wasm-opt` (but obviously without backend).

**Reply by somebody1234** [reply]
can't repro, sorry (i tried on `develop`) - what OS are you using? and does this happen on `develop` too?

**Comment by PabloBuchu** [comment]
@MichaelMauderer can you provide what platform are you working on? 

https://user-images.githubusercontent.com/2855109/231129479-4a6e2eb7-7804-4b26-a5ab-090625793004.mov

I've just tried on my macos and I couldn't reproduce the problem

**Reply by MichaelMauderer** [reply]
I'm on Pop!_OS 22.04 LTS. I can repro this on develop. [This is the log output. ](https://gist.github.com/MichaelMauderer/5d630ba0893e9ca1503d4fc272ee8302). After this, no electron window opens, and the browser cannot connect on `localhost:8080` .

**Reply by somebody1234** [reply]
for comparison [here](https://gist.github.com/somebody1234/5fc1e1b4a39a67a9f8fba3c1027dc33a) are my logs (artix linux, kernel 6.2.9). my setup is a bit weird for some reason though (builds the cli every time) but it should be more or less correct since it works at least?

**Reply by somebody1234** [reply]
for posterity here are the corresponding lines (yours first, mine second)

"Building dynamic assets."
https://gist.github.com/MichaelMauderer/5d630ba0893e9ca1503d4fc272ee8302#file-log-L2663
https://gist.github.com/somebody1234/5fc1e1b4a39a67a9f8fba3c1027dc33a#file-log-L1657

"compressed size of wasm is 16.54MB"
https://gist.github.com/MichaelMauderer/5d630ba0893e9ca1503d4fc272ee8302#file-log-L2668
https://gist.github.com/somebody1234/5fc1e1b4a39a67a9f8fba3c1027dc33a#file-log-L1803

notably, right after, the wasm watcher line right after:
`
	cd "/home/me/git/enso" && "/usr/bin/cargo" "--color" "always" "watch" "--ignore" "README.md" "--" "/home/me/git/enso/target/enso-build/buildscript/enso-build-cli" "--skip-version-check" "--cache-path" "/home/me/.local/share/.enso-ci/cache" "--upload-artifacts" "true" "--repo-path" "/home/me/git/enso" "wasm" "build" "--crate-path" "app/gui" "--wasm-output-path" "/home/me/git/enso/dist/wasm" "--wasm-profile" "release" "--wasm-log-level" "warn" "--wasm-uncollapsed-log-level" "warn" "--skip-wasm-opt" "--wasm-size-limit" "0" "--"`

seems to be missing from your output for some reason.

not that it helps, but `--ignore README.md` seems to be from here (the wasm watcher):

https://github.com/enso-org/enso/blob/87e076720e773ecb0eea6315d1c0c0e71ada5334/build/build/src/project/wasm.rs#L363-L364

which *should* be called from the IDE watcher:
https://github.com/enso-org/enso/blob/87e076720e773ecb0eea6315d1c0c0e71ada5334/build/build/src/project/ide.rs#L150-L151

but i guess not if the artifact is being downloaded from github - not quite sure what determines whether artifacts are built locally or downloaded from github though

**Reply by MichaelMauderer** [reply]
It seems that after checking out develop and doing another clean build, this seems to be working again on my end and I can no longer reproduce this.

**Reply by mwu-tow** [reply]
@somebody1234 
> but i guess not if the artifact is being downloaded from github - not quite sure what determines whether artifacts are built locally or downloaded from github though

This is decided by the `--wasm-source` option. By default, they are built.

**Reply by somebody1234** [reply]
ahh... so that shouldn't affect things here since `--wasm-source` wasn't passed

**Comment by farmaazon** [comment, marked as answer]
@wdanilo is the watch still not working for you? Can we close this thread?

**Reply by wdanilo** [reply]
Thanks Adam, it's working now!
